src/toolbox/shamir_secrets.rs

```diff
@@ -25,7 +94,156 @@ impl SecretShare {
         })
     }
 
+    // given enough shares, output the secret
     pub fn reconstruct(shares: Vec<Scalar>) -> Result<Scalar, String> {
-        return Ok(shares[0]); //TODO Shamir Secret Reconstruction
+        // we need to know t, the threshold for shares
+        // do we assume len(shares)?
+        // I think this is safe, since we're the ones setting up the shares in the first place
+
+        // From Handbook of Applied Cryptography, 12.71:
+        //     S = sum(i=1..t){c_i*y_i}, where c_i = TT(j=1..t, j!=i){x_j / (x_j - x_i)}
+        let mut sum: Scalar = Scalar::zero();
+
+        for i in 0..shares.len() {
+            let y_i = shares[i];
+            // println!("Using point ({}, {:?})", i+1, y_i);
+
+            let mut c_i = 1f64;
+
+            for j in 0..shares.len() {
+                if i == j { continue; }
+                c_i *= ((j+1) as f64) / ((j+1) as f64 - (i+1) as f64);
```

Comment:
Before doing this computation, translate j+1 and i+1 to Scalars, using Scalar::from(), then do this computation as: `Scalar(j+1) * (Scalar(j+1) - Scalar(i+1)).invert()`
You won't need line 122 anymore. (I didn't look at the actual syntax, this is just dummy code)